### DISCUSSION TITLE
crypto is not defined

### DISCUSSION THREAD

**Discussion by evrys** [nuxt/nuxt#19290] [Questions]
I'm trying to get an auth token in a way that's compatible with the Cloudflare Workers environment. CF workers uses the [Web Crypto API](https://developers.cloudflare.com/workers/runtime-apis/web-crypto/). However, in development my API endpoints throw a 500 with `crypto is not defined`.

I am using the `nitro: { preset: "cloudflare" }` configuration. This is an existing project that I'm porting to Nuxt, so I know the code is generally correct; it just doesn't seem to have access to the Web Crypto API in the Nitro dev server context. Ideas appreciated!

**Comment by evrys** [comment]
I've solved it for the moment by conditionally importing a polyfill:

```ts
const webcrypto = typeof crypto !== 'undefined' ? crypto : new (await import("@peculiar/webcrypto")).Crypto
```

**Comment by vis97c** [comment]
This is what I ended up doing.

```js
import { nanoid } from "nanoid";

/**
 * UIID composable
 *
 * @composable
 */
export default function useUIID() {
	try {
		// this should fail in node
		self.crypto || window.crypto;
	} catch (error) {
		// inject crypto into the global node scope
		global.crypto = global.crypto || require("crypto");
	}

	return { uuid: nanoid };
}
```

**Comment by Ak1r0** [comment]
I got a similar problem with the Lucia library using `crypto`. Got this error in dev :

> Reference Error : crypto is not defined

To fix it i've created a Nuxt plugin :
```
# plugins/crypto.ts
export default defineNuxtPlugin(() => {
    if (process.server) {
        const crypto = require('crypto');
        global.crypto = crypto;
    }
})
```

**Reply by w2xi** [reply]
you saved me a morning!  Thanks a lot.

**Reply by KevinLeneschmidt** [reply]
I had the same problem in production using netlify. The default Node.js version was 18x. I updated it to 20x (Site configuration => Build & deploy => Dependency management => Configure => select 20.x), republished the site and everything is okay now!

**Reply by NickBolles** [reply]
fwiw, I added `.server` to the filename to make sure this only runs on the server.

**Reply by Rukhsar** [reply]
You just saved my life, was about to give up on lucia auth.

**Reply by ramijames** [reply]
@KevinLeneschmidt this also was the issue for me. Thanks a lot for the tip!

**Reply by nexdev07** [reply]
I also encountered similar issue, was able to resolve with your help thanks

**1. Install the crypto package**
npm install @peculiar/webcrypto

**2. Create the plugin file**
Create a new file in folder "plugins"
`plugins/polyfill-crypto.ts`

[Contents]
```
export default defineNuxtPlugin(async () => {
  if (process.server && typeof globalThis.crypto === 'undefined') {
    const { Crypto } = await import('@peculiar/webcrypto');
    globalThis.crypto = new Crypto();
  }
});
```
**3. Add Type Declarations (Optional, for TypeScript)**
In types.d.ts (create it in project root if needed):

```
declare global {
  var crypto: Crypto;
  interface Global {
    crypto: Crypto;
  }
}
```

**4. Nuxt Auto-Imports Plugins**
As long as your file is in plugins/, Nuxt will automatically run it. You don’t need to register it manually in nuxt.config.ts.
